futures-util/src/stream/next_some.rs

```diff
@@ -0,0 +1,41 @@
+use core::marker::Unpin;
+use core::pin::Pin;
+use futures_core::stream::{Stream, FusedStream};
+use futures_core::future::{Future, FusedFuture};
+use futures_core::task::{LocalWaker, Poll};
+use crate::stream::StreamExt;
+
+/// A future that resolves to the next value yielded from a [`Stream`].
+#[derive(Debug)]
+#[must_use = "futures do nothing unless polled"]
+pub struct NextSome<'a, St> {
+    stream: Option<&'a mut St>,
+}
+
+impl<'a, St> NextSome<'a, St> {
+    pub(super) fn new(stream: &'a mut St) -> Self {
+        NextSome { stream: Some(stream) }
+    }
+}
+
+impl<'a, St: FusedStream> FusedFuture for NextSome<'a, St> {
+    fn is_terminated(&self) -> bool {
+        self.stream.as_ref().map(|s| s.is_terminated()).unwrap_or(true)
```

Comment:
Hmm, I don't see how.

If the stream has yet to yield `None` then this delegates to the streams `is_terminated()` implementation (this can probably be dropped if it is a contract violation for `FusedStream::is_terminated` to return `true` before yielding `None`, it's just an optimization for streams that can know they're terminated before producing the sentinel element). For a normal stream this will return `false` and the future will go on to be polled.

Once the stream returns `None` then this will return `true`.

Reply:
I guess it's possible to spin loop by constructing multiple of these futures and polling them each once.

That could be avoided by extending the panic to also occur when the stream is terminated, even without polling it.

Reply:
Oh, I understand now. The next time around after they are woken up, the stream will be `None` and the callers of `poll` will get a panic if they're not checking `is_terminated()`. That works, and is a correct implementation of this interface that could be useful in combination with `select!`, but it seems like a dangerous footgun everywhere else, since unless you're checking `is_terminated()` it's basically `await!(s.next()).unwrap()`. This version of `next_some` also doesn't work for the original purpose I added `FuturesUnordered::next_some`, which is special because it can become `Some` again after yielding `None`. I think I'd still be fine w/ adding it just for use w/ `select!`, but we'll have to come up w/ a good name for it to make it clear it should only be used in `select!`, as well as come up w/ a different name for the method on `FuturesUnordered`.

Reply:
Yep, it's definitely just intended for use in `select!` when you don't want to do something special on end-of-stream, and I have no idea of a decent name for it 😦

Reply:
Maybe `select_only_some()`? But finding two different names for this and the one on `FuturesUnordered` is tough.